### PR TITLE
Truncate MCVs

### PR DESCRIPTION
**Pull request by max-hoffman**
Sort and truncate MCVs. Only keep values whose frequency is > twice the uniform frequency. This prevents us from manually summing non-outliers (which is expensive).

### CODE REVIEW

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>0bc232f</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>0bc232f</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_p95 | to_latency_p95 | is_faster |
|---------------------|------------------|----------------|-----------|
| tpcc-scale-factor-1 | 74.46            | 74.46          | 0         |


|      test_name      | server_name |              server_version              |  tps  |      test_name      | server_name |              server_version              |  tps  | is_faster |
|---------------------|-------------|------------------------------------------|-------|---------------------|-------------|------------------------------------------|-------|-----------|
| tpcc-scale-factor-1 | dolt        | b4dc47360dc0d5235634c3e5e219110d5d1a1ddf | 33.54 | tpcc-scale-factor-1 | dolt        | 0bc232f5392a6fed47f6f7d6cc3d143b428cdc79 | 32.85 | 0         |

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>39bcc83</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>39bcc83</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 2.81                | 2.81              | 0         |
| groupby_scan         | 17.01               | 17.32             | 0         |
| index_join           | 5.28                | 5.28              | 0         |
| index_join_scan      | 3.25                | 2.52              | 1         |
| index_scan           | 53.85               | 53.85             | 0         |
| oltp_point_select    | 0.46                | 0.46              | 0         |
| oltp_read_only       | 7.56                | 7.56              | 0         |
| select_random_points | 0.75                | 0.75              | 0         |
| select_random_ranges | 0.9                 | 0.9               | 0         |
| table_scan           | 54.83               | 54.83             | 0         |
| types_table_scan     | 139.85              | 139.85            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 6.09                | 6.09              | 0         |
| oltp_insert           | 3.02                | 3.02              | 0         |
| oltp_read_write       | 13.95               | 13.95             | 0         |
| oltp_update_index     | 3.07                | 3.07              | 0         |
| oltp_update_non_index | 3.02                | 3.02              | 0         |
| oltp_write_only       | 6.43                | 6.43              | 0         |
| types_delete_insert   | 6.67                | 6.67              | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>c9875b3</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>c9875b3</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>ba8e7f4</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>ba8e7f4</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Review by jycor — APPROVED** [review]
LGTM

File: go/libraries/doltcore/sqle/statspro/update.go [hunk 1]
```diff
@@ -365,6 +370,21 @@ func (m mcvHeap) Counts() []uint64 {
...
 
+func (m mcvHeap) Sort() {
+	sort.Slice(m, m.Less)
+}
+
+func (m *mcvHeap) Truncate(cutoff float64) {
+	start := m.Len()
+	for i, v := range *m {
+		if float64(v.cnt) >= cutoff {
+			start = i
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/statspro/update.go:381]: This should break out of the loop here, right?

File: go/libraries/doltcore/sqle/statspro/update_test.go [hunk 2]
```diff
@@ -165,12 +165,25 @@ func TestBucketBuilder(t *testing.T) {
...
-				McvVals:     []sql.Row{{"i", int64(1)}, {"g", int64(2)}, {"h", int64(1)}},
-				McvsCnt:     []uint64{2, 2, 2},
+				McvVals:     []sql.Row{},
+				McvsCnt:     []uint64{},
 				BoundVal:    sql.Row{"i", int64(1)},
 				BoundCnt:    2,
 			}},
 		},
+		{
+			name:    "mcvs",
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/statspro/update_test.go:175]: Lets also add a test with multiple mcvs.

File: integration-tests/bats/stats.bats [hunk 3]
```diff
@@ -208,33 +208,13 @@ teardown() {
     cd repo2
 
     dolt sql -q "alter table xy add index y2 (y)"
-    dolt sql -q "insert into xy values (0,0), (1,0), (2,0), (3,0), (4,0), (5,0)"
+    dolt sql -q "insert into xy values (0,0), (1,0), (2,0), (3,0), (4,0), (5,0), (6,1), (7,2), (8,3), (9,4)"
 
-    # setting variables doesn't hang or error
-    dolt sql -q "SET @@persist.dolt_stats_auto_refresh_enabled = 1;"
```
Comment (nicktobey) [inline integration-tests/bats/stats.bats:214]: Why were these deleted?

**Review by nicktobey — APPROVED** [review]
A bunch of tests now return no mcv information. I'd make sure that we're not losing any test coverage as a result. Otherwise, this looks good.

File: integration-tests/bats/stats.bats (hunk 3 above)
Comment (max-hoffman) [reply integration-tests/bats/stats.bats:214]: replaced with the analyze below, we already test background loading figured this was overkill

File: go/libraries/doltcore/sqle/statspro/update.go (hunk 1 above)
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/statspro/update.go:381]: follow up bug fixes here https://github.com/dolthub/dolt/pull/8049
